Cloudwatch: Fix bug with opening dashboards with over 50 log group names

sarahzinger opened the pull request:
**What this PR does / why we need it**:
Right now if you have more than 50 cloudwatch log group names, make a dashboard, save that dashboard and then try to edit it later you'll run into some bugs. The main gist is that we attempt to smartly clear the log group names if the log group names are no longer valid options, and we are incorrectly deciding that if a log group name does not return in the initial fetch it must be an invalid name. 

[One proposed community solution](https://github.com/grafana/grafana/pull/56185) was to instead fetch all group names when we load the component. This is likely to be a very effective solution for many use cases, but it does seem likely that it won't scale well for accounts that have say thousands of log groups. However there were additional tests in that pr which we should probably consider cherry picking! 

This proposed solution will instead not clear the log group names, even if they are no longer valid (ex: you change regions). While it is a bit less slick from a ui perspective, I think ultimately what we thought was a feature (to remove log group names when regions are changed) may actually be a bug and pain point, as uses may have log groups named the same thing from region to region and not want to have to reselect it every time. 

Another factor here is that Log Group names are selected on the Config Page in Cloudwatch. This adds its own complexity in the code to determine if changes have been made to the connection details and whether or not the cloudwatch logs selection should reset or not. We also used to force a save on open of the log group names, so as to ensure we had up to date credential information to which to make a request for log groups. In the interest of trying to simplify the code and make the user experience more predictable, I'm proposing adding a "save" button after connection config that is disabled by default and then enabled when credentials change, indicating to the user that they need to save these configuration settings before making future selections like default log groups. This allows us to move any logic around "saved" vs not saved out of log group selection, and keep that in the config component. It does add one more button for users to have to click, but I'm hoping it is more intuitive than saving when opening the log group names selection.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/55700

**Special notes for your reviewer**:
- I started writing some tests but not as many as I'd like. But I also feel like this is a fairly painful bug for users that we should fix sooner rather than later so it might be best to merge with what we've got (assuming we're ok with the solution!) and make a ticket to add some more tests later.

sarahzinger committed:
Fix 55700: issue with Cloudwatch log group names

fridgepoet commented:
Thanks a lot for the clear description and background of the problem.
Understood this might have taken some time --making a suggestion here that may not be valid as I don't have all the changes in my head as you do.

I'm seeing two distinct changes here:
* not clear the log group names
* adding a "save" button after connection config that is disabled by default and then enabled when credentials change

What do you think about making separate PRs for those two parts? 

Also let us know how we can help out for test writing --I would feel much more confident merging this work with the tests needed to protect, ensure, and document its changes. Breaking up this PR into smaller parts may help us to tackle that work together if you'd like. 
Were some of those tests from the community PR something we could have as co-authored changes from them?

sarahzinger commented:
Hi @fridgepoet great questions!

Previously the way the LogGroupSelector worked is that it would refetch log groups whenever the datasource, region or connection details changed: https://github.com/grafana/grafana/pull/57081/files#diff-ba51f30f7aa21b34e9220218ecb799d788e46ccaa3ddac87101f9636d7c5648eL101

This was causing the originally buggy behavior of clearing the old log group names. So I changed it so that we would refetch log groups when users open the modal instead. 

However previously when connection details were updated, we would save the new credentials on menu open: https://github.com/grafana/grafana/pull/57081/files#diff-4857b13d454eb8d6952e49135e8cae24f0c268f1155111caae18eae3baded193L117 and so when I added a fetch after a save I ran into a race condition where we were fetching the log groups but the new connection details hadn't saved yet. 

So if I keep all of these changes but remove the save button users will fill out connection details (such as changing an access key) and then attempt to open log groups and receive nothing or incorrect details back because it's not clear they need to save their credentials again before fetching the log groups. 

Regarding testing: I was having difficulty testing our Select components, I find them to be pretty finicky with react-testing-library. The tests I mostly wanted to add were in the ConfigEditor.test.tsx to verify saving logic of the new button. But kept running into issues. If you we feel very strongly about it I can take more time to add tests, but I'm worried about user impact here, and I think ultimately we have plans to work on this cloudwatch log selector in future projects so I think there will be additional opportunities to revisit. But let me know if you disagree! It is always unpleasant when code coverage scores go down lol, possibly I'm being too nonchalant.
